### PR TITLE
[Model Runner V2] Apply synthetic mode to probabilistic rejection sampler

### PR DESCRIPTION
**Pull request by TheEpicDolphin**
# Context
Per-position acceptance rates were recently added in [this PR](https://github.com/vllm-project/vllm/pull/40662), adding support for both MRV1 and MRV2. However, MRV2 currently uses [a separate path](https://github.com/vllm-project/vllm/blob/a0dc7a0f368a3f212d37e39ba4dd7b682e378fe2/vllm/v1/worker/gpu/spec_decode/rejection_sampler.py#L140-L152) for synthetic rejection sampling, which is not ideal because synthetic acceptance rates should closely approximate actual rejection sampling.

# This PR
I apply the synthetic mode in the probabilistic rejection sampling kernels. This involved pass the synthetic conditional rates to `_probabilistic_rejection_kernel` (now renamed to `_rejection_kernel`). If non-None, the draft token is accepted with the conditional probability of that given step. Note that the kernel still loads target/draft logits, sum-exponentials, etc, even if unused to closely match the performance of actual rejection sampling.

Other notable changes:
- Deleted `synthetic_rejection_sampler_utils.py`, not used anymore.
- Renamed `probabilistic_rejection_sampler_utils.py` to `rejection_sampler_utils.py`, as well as removed `probabilistic` prefix from all methods within. That prefix was leftover from when the rejection sample method was either "probabilistic", "strict", or "synthetic". Now it's just either "standard" or "synthetic".
- Added test cases for synthetic rejection sampling to `test_rejection_sampler_utils.py`.

# Tests
## Automatic
Added `test_synthetic_rejection_sample` test to `test_rejection_sampler_utils.py`, that verifies that the number of accepted tokens approximately equals the per-position acceptance rates over many trials. This is verified for both greedy and stochastic requests.
```
(vllm-1) gdelfin@h200-1:~/vllm-1$ python -m pytest tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample -v
=================================================================== test session starts ====================================================================
.
.
.
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-1.0-unconditional_rates0] PASSED                             [ 12%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-0.0-unconditional_rates1] PASSED                             [ 25%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-1.0-unconditional_rates2] PASSED                             [ 37%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-0.0-unconditional_rates3] PASSED                             [ 50%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-1.0-unconditional_rates4] PASSED                             [ 62%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[3-0.0-unconditional_rates5] PASSED                             [ 75%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[1-1.0-unconditional_rates6] PASSED                             [ 87%]
tests/v1/spec_decode/test_rejection_sampler_utils.py::test_synthetic_rejection_sample[1-0.0-unconditional_rates7] PASSED                             [100%]

==================================================================== 8 passed in 4.98s =====================================================================
```

## Manual
For manual testing, I ran a server with **standard rejection sampling**:
```
VLLM_USE_V2_MODEL_RUNNER=1 vllm serve meta-llama/Meta-Llama-3-8B-Instruct \
  --no-enable-prefix-caching \
  --tensor-parallel-size 1 \
  --data-parallel-size 1 \
  --speculative-config '{"method": "eagle", "model": "yuhuili/EAGLE-LLaMA3-Instruct-8B", "num_speculative_tokens": 3, "draft_sample_method": "gumbel", "rejection_sample_method": "standard"}'
```

And benchmarked it with the following command:
```
vllm-bench \
  --backend openai-chat \
  --base-url http://127.0.0.1:8000 \
  --model meta-llama/Meta-Llama-3-8B-Instruct \
  --dataset-name hf \
  --dataset-path philschmid/mt-bench \
  --ignore-eos \
  --request-rate inf \
  --temperature 0.0 \
  --num-warmups 50 \
  --output-len 1536 \
  --max-concurrency 64 \
  --num-prompts 640
```

Then, I took the acceptance rates reported in the stats and set them as the synthetic acceptance rates for a new server instance:
```
VLLM_USE_V2_MODEL_RUNNER=1 vllm serve meta-llama/Meta-Llama-3-8B-Instruct \
  --no-enable-prefix-caching \
  --tensor-parallel-size 1 \
  --data-parallel-size 1 \
  --speculative-config '{"method": "eagle", "model": "yuhuili/EAGLE-LLaMA3-Instruct-8B", "num_speculative_tokens": 3, "draft_sample_method": "gumbel", "rejection_sample_method": "synthetic", "synthetic_acceptance_rates": <reported_acceptance_rates>}'
```
And ran the same benchmark on it. Below are the observed results for temperatures 0 and 1:

### Temperature = 0.0
| Metric | Actual | Synthetic | Delta |
|--------|-------:|----------:|------:|
| **General** | | | |
| Successful requests | 640 | 640 | 0 |
| Failed requests | 0 | 0 | 0 |
| Maximum request concurrency | 64 | 64 | 0 |
| Benchmark duration (s) | 8.79 | 8.56 | -2.6% |
| Total input tokens | 42596 | 42596 | 0 |
| Total generated tokens | 81920 | 81920 | 0 |
| Request throughput (req/s) | 72.79 | 74.81 | +2.8% |
| Output token throughput (tok/s) | 9316.55 | 9575.57 | +2.8% |
| Peak output token throughput (tok/s) | 5043.00 | 5183.00 | +2.8% |
| Peak concurrent requests | 147.00 | 151.00 | +2.7% |
| Total token throughput (tok/s) | 14160.88 | 14554.59 | +2.8% |
| **Time to First Token** | | | |
| Mean TTFT (ms) | 74.52 | 59.99 | -19.5% |
| Median TTFT (ms) | 43.31 | 46.23 | +6.7% |
| P99 TTFT (ms) | 306.33 | 203.52 | -33.6% |
| P90 TTFT (ms) | 240.57 | 104.34 | -56.6% |
| **Time per Output Token (excl. 1st token)** | | | |
| Mean TPOT (ms) | 6.07 | 6.10 | +0.5% |
| Median TPOT (ms) | 5.97 | 6.14 | +2.8% |
| P99 TPOT (ms) | 8.59 | 7.36 | -14.3% |
| P90 TPOT (ms) | 7.55 | 6.84 | -9.4% |
| **Inter-token Latency** | | | |
| Mean ITL (ms) | 12.95 | 11.87 | -8.3% |
| Median ITL (ms) | 11.21 | 10.71 | -4.5% |
| P99 ITL (ms) | 44.48 | 31.53 | -29.1% |
| P90 ITL (ms) | 16.47 | 15.58 | -5.4% |
| **End-to-end Latency** | | | |
| Mean E2EL (ms) | 845.61 | 835.01 | -1.3% |
| Median E2EL (ms) | 831.82 | 833.47 | +0.2% |
| P99 E2EL (ms) | 1197.71 | 1012.59 | -15.5% |
| P90 E2EL (ms) | 1039.80 | 932.29 | -10.3% |
| **Speculative Decoding** | | | |
| Acceptance rate (%) | 32.98 | 32.62 | -0.36pp |
| Acceptance length | 1.99 | 1.98 | -0.5% |
| Drafts | 41096 | 41332 | +0.6% |
| Draft tokens | 123288 | 123996 | +0.6% |
| Accepted tokens | 40659 | 40447 | -0.5% |
| Position 0 acceptance (%) | 56.40 | 55.84 | -0.56pp |
| Position 1 acceptance (%) | 28.46 | 28.21 | -0.25pp |
| Position 2 acceptance (%) | 14.08 | 13.81 | -0.27pp |

### Temperature = 1.0
| Metric | Actual | Synthetic | Delta |
|--------|-------:|----------:|------:|
| **General** | | | |
| Successful requests | 640 | 640 | 0 |
| Failed requests | 0 | 0 | 0 |
| Maximum request concurrency | 64 | 64 | 0 |
| Benchmark duration (s) | 9.86 | 9.98 | +1.2% |
| Total input tokens | 42596 | 42596 | 0 |
| Total generated tokens | 81920 | 81920 | 0 |
| Request throughput (req/s) | 64.89 | 64.11 | -1.2% |
| Output token throughput (tok/s) | 8306.49 | 8205.90 | -1.2% |
| Peak output token throughput (tok/s) | 4730.00 | 4702.00 | -0.6% |
| Peak concurrent requests | 137.00 | 139.00 | +1.5% |
| Total token throughput (tok/s) | 12625.62 | 12472.73 | -1.2% |
| **Time to First Token** | | | |
| Mean TTFT (ms) | 70.78 | 62.77 | -11.3% |
| Median TTFT (ms) | 44.08 | 49.26 | +11.7% |
| P99 TTFT (ms) | 281.72 | 198.31 | -29.6% |
| P90 TTFT (ms) | 141.36 | 101.85 | -28.0% |
| **Time per Output Token (excl. 1st token)** | | | |
| Mean TPOT (ms) | 6.97 | 7.17 | +2.9% |
| Median TPOT (ms) | 6.90 | 7.14 | +3.5% |
| P99 TPOT (ms) | 10.21 | 8.70 | -14.8% |
| P90 TPOT (ms) | 8.66 | 7.97 | -8.0% |
| **Inter-token Latency** | | | |
| Mean ITL (ms) | 14.54 | 13.30 | -8.5% |
| Median ITL (ms) | 12.36 | 12.33 | -0.2% |
| P99 ITL (ms) | 62.92 | 37.91 | -39.7% |
| P90 ITL (ms) | 16.75 | 16.88 | +0.8% |
| **End-to-end Latency** | | | |
| Mean E2EL (ms) | 956.23 | 972.96 | +1.8% |
| Median E2EL (ms) | 941.24 | 973.66 | +3.4% |
| P99 E2EL (ms) | 1362.13 | 1197.84 | -12.1% |
| P90 E2EL (ms) | 1172.02 | 1082.24 | -7.7% |
| **Speculative Decoding** | | | |
| Acceptance rate (%) | 29.83 | 29.76 | -0.07pp |
| Acceptance length | 1.89 | 1.89 | 0.0% |
| Drafts | 43171 | 43182 | +0.0% |
| Draft tokens | 129513 | 129546 | +0.0% |
| Accepted tokens | 38629 | 38551 | -0.2% |
| Position 0 acceptance (%) | 52.58 | 52.60 | +0.02pp |
| Position 1 acceptance (%) | 25.15 | 25.07 | -0.08pp |
| Position 2 acceptance (%) | 11.75 | 11.60 | -0.15pp |

The reported per-position acceptance rates closely match, as do the other performance statistics, such as TPOT, output token throughput, etc. This demonstrates that synthetic rejection sampling is producing correct per-position acceptance rates, and can be used to simulate rejection sampling.

## Accuracy Benchmark
**Server Command**
```
VLLM_USE_V2_MODEL_RUNNER=1 vllm serve meta-llama/Meta-Llama-3-8B-Instruct \
  --no-enable-prefix-caching \
  --tensor-parallel-size 1 \
  --data-parallel-size 1 \
  --speculative-config '{"method": "eagle", "model": "yuhuili/EAGLE-LLaMA3-Instruct-8B", "num_speculative_tokens": 3, "rejection_sample_method": "standard"}'
```
I benchmarked using **gsm8k** and observed the following results:
| Temperature | Metric | Main | This PR | Delta |
|:-----------:|--------|:----:|:------:|:-----:|
| 0.0 | flexible-extract | 0.7445 | 0.7422 | -0.23% |
| 0.0 | strict-match | 0.7445 | 0.7453 | +0.08% |
| 0.6 | flexible-extract | 0.7483 | 0.7559 | +0.76% |
| 0.6 | strict-match | 0.7513 | 0.7551 | +0.38% |

No regressions from this PR.

### CODE REVIEW

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py [hunk 1]
```diff
@@ -275,9 +288,14 @@ def _probabilistic_rejection_kernel(
                 else:
                     # One-hot draft: q(draft_token) = 1, log_q = 0.
                     draft_log_prob = 0
-                # Probability ratio test: p(x) > u * q(x)
-                # Equivalent log form: log_p(x) > log(u) + log_q(x)
-                accepted &= target_log_prob > tl.log(u) + draft_log_prob
+
+                if SYNTHETIC_MODE:
```
[the remark this answers is not in the document]
Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py:292]: The LSE and log-probability calculations are necessary to match performance of standard rejection sampling.

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py [hunk 2]
```diff
@@ -489,6 +507,8 @@ def probabilistic_rejection_sample(
     temperature: torch.Tensor,
     # [max_num_reqs]
     seed: torch.Tensor,
+    # [num_speculative_steps]
+    synthetic_conditional_rates: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py]: Addressed!

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py [hunk 3]
```diff
@@ -198,9 +198,12 @@ def _probabilistic_rejection_kernel(
     seed_ptr,
     # [num_logits]
     pos_ptr,
+    # [num_speculative_steps]
+    synthetic_conditional_rates_ptr,
     vocab_num_blocks,
     PADDED_VOCAB_NUM_BLOCKS: tl.constexpr,
     HAS_DRAFT_LOGITS: tl.constexpr,
+    SYNTHETIC_MODE: tl.constexpr,
```
[the remark this answers is not in the document]
Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py:206]: The residual distribution logic is needed to match performance of standard rejection sampling. This was an intentional choice.

**Review by benchislett — APPROVED** [review]
LGTM. While I appreciate the thorough evaluation, I feel like a further test with GSM8k + spec decoding (not synthetic) is warranted just to make sure we're checking fundamental specdec correctness.

**Comment by TheEpicDolphin** [comment]
Added GSM8k + spec decoding accuracy benchmark!
